File: services/pkgd/tests/support/apps/create-app.ts
import {
  Metadata
} from "@grpc/grpc-js";
import type {
  App,
  CreateAppRequest,
  PackageServiceClient
} from "../../generated/v1/pkgd.js";
import {
  callUnary
} from "../call-unary.js";

export async function createApp(
  client: PackageServiceClient,
  request: CreateAppRequest,
  metadata = new Metadata()
): Promise<App> {
  return await callUnary<App>((done) =>
    client.createApp(request, metadata, done));
}

File: services/pkgd/csharp/src/CtlFlow.Packages.Pkgd.Service/Grpc/Apps/CreateApp.cs
using System.Diagnostics;
using CtlFlow.Packages.V1;
using Grpc.Core;
using static CtlFlow.Packages.Pkgd.Service.Auditing.AuditDelivery;
using static CtlFlow.Packages.Pkgd.Service.Authorization.PackageAuthorization;
using static CtlFlow.Packages.Pkgd.Service.Grpc.Requests.PackageRequests;
using AppDatabase = CtlFlow.Packages.Pkgd.Db.Apps.Apps;

namespace CtlFlow.Packages.Pkgd.Service.Grpc;

internal sealed partial class PackageGrpcService
{
    public override async Task<App> CreateApp(
        CreateAppRequest request,
        ServerCallContext context)
    {
        var identity = await AuthenticateAppCreation(context);
        var draft = await CreateAppDraft(
            request,
            context.CancellationToken);
        await AuthorizeCapability(
            _policyClient,
            _settings.Policy,
            _telemetry,
            identity,
            Authorization.PkgdCapability.CreateApp,
            draft.Scope,
            null,
            context.CancellationToken);
        var audit = await CreateAuditContext(
            identity,
            Activity.Current,
            DateTimeOffset.UtcNow,
            context.CancellationToken);
        var result = await AppDatabase.CreateApp(
            _packageDatabase,
            draft,
            audit,
            context.CancellationToken);
        return await CompleteAppMutation(
            result,
            context.CancellationToken);
    }
}
